test/Container/ApplicationFactoryTest.php

```diff
@@ -200,4 +200,489 @@ public function testWillUseSaneDefaultsForOptionalServices()
         $this->assertInstanceOf('Zend\Diactoros\Response\SapiEmitter', $app->getEmitter()->pop());
         $this->assertInstanceOf('Zend\Stratigility\FinalHandler', $app->getFinalHandler());
     }
+
+    /**
+     * @group piping
+     */
+    public function testCanPipeMiddlewareProvidedDuringConfigurationPriorToSettingRoutes()
+    {
+        $router = $this->prophesize('Zend\Expressive\Router\RouterInterface');
+        $emitter = $this->prophesize('Zend\Diactoros\Response\EmitterInterface');
+        $finalHandler = function ($req, $res, $err = null) {
+        };
+
+        $middleware = function ($req, $res, $next = null) {
+        };
+
+        $config = [
+            'routes' => [
+                [
+                    'path' => '/',
+                    'middleware' => 'HelloWorld',
+                    'allowed_methods' => [ 'GET' ],
+                ],
+            ],
+            'middleware_pipeline' => [
+                'pre_routing' => [
+                    [ 'middleware' => $middleware ],
+                    [ 'path' => '/foo', 'middleware' => $middleware ],
+                ],
+                'post_routing' => [ ],
+            ],
+        ];
+
+        $this->container
+            ->has('Zend\Expressive\Router\RouterInterface')
+            ->willReturn(true);
+        $this->container
+            ->get('Zend\Expressive\Router\RouterInterface')
+            ->will(function () use ($router) {
+                return $router->reveal();
+            });
+
+        $this->container
+            ->has('Zend\Diactoros\Response\EmitterInterface')
+            ->willReturn(true);
+        $this->container
+            ->get('Zend\Diactoros\Response\EmitterInterface')
+            ->will(function () use ($emitter) {
+                return $emitter->reveal();
+            });
+
+        $this->container
+            ->has('Zend\Expressive\FinalHandler')
```

Comment:
I suggest leaving as-is until I have a decision on #18. If we decide not to make that change, I don't want to have to revert, and there are reasons we may not make that change due to planned features.